File: platforms/mate/applet.py
#!/usr/bin/env python3
"""Native MATE panel host for the UsageStat service (GTK 3, out of process)."""
import json
import gi
gi.require_version('Gtk', '3.0')
gi.require_version('MatePanelApplet', '4.0')
gi.require_foreign('cairo')
from gi.repository import Gtk, Gdk, GdkPixbuf, Gio, GLib, MatePanelApplet

BUS = 'io.github.HashimK.UsageStatBar'
PATH = '/io/github/HashimK/UsageStatBar'
IFACE = BUS + '1'

class Indicator:
    def __init__(self, applet):
        self.applet = applet
        self.bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
        self.cancellable = Gio.Cancellable()
        self.state = None
        self.scroll_amount = 0
        self.scroll_time = 0
        self.image = Gtk.Image.new_from_icon_name('office-chart-pie', Gtk.IconSize.MENU)
        self.button = Gtk.Button()
        self.button.set_relief(Gtk.ReliefStyle.NONE)
        self.button.set_name('usagestat-panel')
        css = Gtk.CssProvider()
        css.load_from_data(b'#usagestat-panel { padding: 0 4px; border: none; box-shadow: none; background: transparent; color: inherit; }'
                           b'#usagestat-panel:hover { background: alpha(currentColor, 0.08); border-radius: 6px; }')
        self.button.get_style_context().add_provider(css, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION)
        self.button.add(self.image)
        self.button.connect('clicked', self.clicked)
        self.button.add_events(Gdk.EventMask.SCROLL_MASK | Gdk.EventMask.SMOOTH_SCROLL_MASK | Gdk.EventMask.BUTTON_PRESS_MASK)
        self.button.connect('scroll-event', self.scroll)
        self.button.connect('button-press-event', self.press)
        self.actions = Gtk.ActionGroup(name='UsageStatActions')
        preferences = Gtk.Action(name='UsageStatPreferences', label='UsageStat preferences', tooltip='Configure UsageStat', stock_id=None)
        preferences.connect('activate', lambda *_: self.call('Preferences', '(s)', ('',)))
        self.actions.add_action(preferences)
        applet.setup_menu('<menuitem name="UsageStatPreferences" action="UsageStatPreferences"/>', self.actions)
        self.button.connect('style-updated', lambda *_: self.redraw())
        self.image.connect('notify::scale-factor', lambda *_: self.redraw())
        applet.add(self.button)
        applet.set_flags(MatePanelApplet.AppletFlags.EXPAND_MINOR)
        applet.connect('destroy', self.close)
        applet.connect('change-size', lambda *_: self.redraw())
        applet.connect('change-orient', lambda *_: self.redraw())
        self.signal = self.bus.signal_subscribe(BUS, IFACE, 'Changed', PATH, None, Gio.DBusSignalFlags.NONE,
            lambda _bus, _sender, _path, _iface, _signal, args: self.render(json.loads(args.unpack()[0])))
        self.watch = Gio.bus_watch_name(Gio.BusType.SESSION, BUS, Gio.BusNameWatcherFlags.AUTO_START,
            lambda *_: self.request(), lambda *_: self.button.set_tooltip_text('UsageStat is stopped'))
        applet.show_all()

    def call(self, method, signature=None, args=None, done=None):
        def finished(bus, result):
            try:
                values = bus.call_finish(result).unpack()
                if done and not self.cancellable.is_cancelled(): done(values)
            except GLib.Error as error:
                if not self.cancellable.is_cancelled(): self.button.set_tooltip_text(error.message)
        self.bus.call(BUS, PATH, IFACE, method, GLib.Variant(signature, args) if signature else None,
            None, Gio.DBusCallFlags.NONE, 5000, self.cancellable, finished)

    def request(self):
        self.call('GetSnapshot', done=lambda value: self.render(json.loads(value[0])))

    def clicked(self, *_):
        top = self.button.get_toplevel()
        window = top.get_window()
        _, x, y = window.get_origin()
        coordinates = self.button.translate_coordinates(top, 0, 0)
        # PyGObject's GTK override returns (x, y), or None on failure. Older
        # unoverridden bindings also include the success boolean.
        if coordinates is None or (len(coordinates) == 3 and not coordinates[0]):
            return self.call('ToggleDetails', '(s)', ('',))
        dx, dy = coordinates[-2:]
        monitor = self.button.get_display().get_monitor_at_window(window)
        area, screen = monitor.get_workarea(), monitor.get_geometry()
        vertical = self.applet.get_orient() in [MatePanelApplet.AppletOrient.LEFT, MatePanelApplet.AppletOrient.RIGHT]
        edge = ('left' if x < screen.x + screen.width/2 else 'right') if vertical else ('top' if y < screen.y + screen.height/2 else 'bottom')
        scale = window.get_scale_factor()
        rect = lambda x, y, w, h: dict(x=x*scale, y=y*scale, w=w*scale, h=h*scale)
        anchor = dict(edge=edge, rect=rect(x+dx, y+dy, self.button.get_allocated_width(), self.button.get_allocated_height()),
                      work=rect(area.x, area.y, area.width, area.height))
        self.call('ToggleDetailsAt', '(ss)', ('', json.dumps(anchor)))

    def render(self, state):
        if self.cancellable.is_cancelled(): return
        self.state = state
        self.redraw()
        self.button.set_tooltip_text('\n'.join(f"{p['name']}: {p['error'] or p['text']}" for p in state['providers'] if not p['parent']))

    def redraw(self):
        if not self.state or self.cancellable.is_cancelled(): return
        foreground = self.button.get_style_context().get_color(Gtk.StateFlags.NORMAL)
        vertical = self.applet.get_orient() in [MatePanelApplet.AppletOrient.LEFT, MatePanelApplet.AppletOrient.RIGHT]
        light = (foreground.red + foreground.green + foreground.blue) < 1.5
        key = ('panelImageVerticalLight' if light else 'panelImageVertical') if vertical else ('panelImageLight' if light else 'panelImage')
        path = self.state.get(key + 'Png', self.state[key])
        height = max(16, min(40 if vertical else 28, self.applet.get_size() - 4))
        scale = self.image.get_scale_factor()
        try:
            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_scale(path, height * scale if vertical else -1, -1 if vertical else height * scale, True)
            self.image.set_from_surface(Gdk.cairo_surface_create_from_pixbuf(pixbuf, scale, None))
        except GLib.Error as error:
            self.button.set_tooltip_text(error.message)

    def scroll(self, _button, event):
        if self.state and not self.state.get('interaction', {}).get('panelScroll', True): return False
        if event.direction == Gdk.ScrollDirection.SMOOTH:
            valid, dx, dy = event.get_scroll_deltas()
            if not valid: return False
            delta = dy if abs(dy) >= abs(dx) else dx
            now = GLib.get_monotonic_time()
            if now - self.scroll_time > 250000 or delta * self.scroll_amount < 0: self.scroll_amount = 0
            self.scroll_time = now
            self.scroll_amount += delta
            if abs(self.scroll_amount) >= 1:
                self.call('Scroll', '(i)', (1 if self.scroll_amount > 0 else -1,))
                self.scroll_amount = 0
            return True
        if event.direction in [Gdk.ScrollDirection.UP, Gdk.ScrollDirection.DOWN, Gdk.ScrollDirection.LEFT, Gdk.ScrollDirection.RIGHT]:
            self.scroll_amount = 0
            self.call('Scroll', '(i)', (-1 if event.direction in [Gdk.ScrollDirection.UP, Gdk.ScrollDirection.LEFT] else 1,))
            return True
        return False

    def press(self, _button, event):
        # Let MATE provide its native Move / Lock / Remove context menu.
        # UsageStat preferences are added to that menu above.
        if event.button == 2:
            self.call('Refresh')
            return True
        return False

    def close(self, *_):
        self.cancellable.cancel()
        self.bus.signal_unsubscribe(self.signal)
        Gio.bus_unwatch_name(self.watch)

def factory(applet, iid, _data):
    if iid != 'UsageStatApplet': return False
    applet._usage_stat = Indicator(applet)
    return True

MatePanelApplet.Applet.factory_main('UsageStatAppletFactory', True, MatePanelApplet.Applet.__gtype__, factory, None)

File: platforms/cinnamon/applet.js
const Applet = imports.ui.applet;
const PopupMenu = imports.ui.popupMenu;
const Main = imports.ui.main;
const Gio = imports.gi.Gio;
const GLib = imports.gi.GLib;
const St = imports.gi.St;
const Clutter = imports.gi.Clutter;
const Cogl = imports.gi.Cogl;
const GdkPixbuf = imports.gi.GdkPixbuf;
const BUS = 'io.github.HashimK.UsageStatBar';
const PATH = '/io/github/HashimK/UsageStatBar';
const IFACE = 'io.github.HashimK.UsageStatBar1';

class UsageStatApplet extends Applet.Applet {
    constructor(metadata, orientation, height, instanceId) {
        super(orientation, height, instanceId);
        this._closed = false;
        this._orientation = orientation;
        this._scrollAmount = 0;
        this._scrollTime = 0;
        this._cancellable = new Gio.Cancellable();
        this._image = new St.Widget({width: 150, height: 28});
        this.actor.add_child(this._image);
        this.set_applet_tooltip('UsageStat Bar — starting');
        this.setAllowedLayout(Applet.AllowedLayout.BOTH);
        this._applet_context_menu.addMenuItem(new PopupMenu.PopupMenuItem('Refresh usage'));
        this._applet_context_menu._getMenuItems().slice(-1)[0].connect('activate', () => this._call('Refresh'));
        const prefs = new PopupMenu.PopupMenuItem('UsageStat preferences');
        prefs.connect('activate', () => this._call('Preferences', '(s)', ''));
        this._applet_context_menu.addMenuItem(prefs);
        this.actor.connect('scroll-event', (_actor, event) => {
            if (this._state?.interaction?.panelScroll === false) return Clutter.EVENT_PROPAGATE;
            const direction = event.get_scroll_direction();
            if (direction === Clutter.ScrollDirection.SMOOTH) {
                const [dx, dy] = event.get_scroll_delta();
                const delta = Math.abs(dy) >= Math.abs(dx) ? dy : dx;
                const now = GLib.get_monotonic_time();
                if (now - this._scrollTime > 250000 || delta * this._scrollAmount < 0) this._scrollAmount = 0;
                this._scrollTime = now;
                this._scrollAmount += delta;
                if (Math.abs(this._scrollAmount) >= 1) {
                    this._call('Scroll', '(i)', this._scrollAmount > 0 ? 1 : -1);
                    this._scrollAmount = 0;
                }
                return Clutter.EVENT_STOP;
            }
            if ([Clutter.ScrollDirection.UP, Clutter.ScrollDirection.DOWN, Clutter.ScrollDirection.LEFT, Clutter.ScrollDirection.RIGHT].includes(direction)) {
                this._scrollAmount = 0;
                this._call('Scroll', '(i)', [Clutter.ScrollDirection.UP, Clutter.ScrollDirection.LEFT].includes(direction) ? -1 : 1);
                return Clutter.EVENT_STOP;
            }
            return Clutter.EVENT_PROPAGATE;
        });
        this._subscription = Gio.DBus.session.signal_subscribe(BUS, IFACE, 'Changed', PATH, null, Gio.DBusSignalFlags.NONE,
            (_bus, _sender, _path, _iface, _signal, params) => this._render(JSON.parse(params.deep_unpack()[0])));
        this.actor.connect('style-changed', () => { if (this._state) this._render(this._state); });
        this._theme = St.ThemeContext.get_for_stage(global.stage);
        this._scaleSignal = this._theme.connect('notify::scale-factor', () => { if (this._state) this._render(this._state); });
        this._watch = Gio.bus_watch_name(Gio.BusType.SESSION, BUS, Gio.BusNameWatcherFlags.AUTO_START,
            () => this._call('GetSnapshot', null, null, value => this._render(JSON.parse(value[0]))),
            () => { if (!this._closed) this.set_applet_tooltip('UsageStat is stopped. Click to start it.'); });
    }
    _call(method, signature = null, value = null, done = null) {
        Gio.DBus.session.call(BUS, PATH, IFACE, method, signature ? new GLib.Variant(signature, Array.isArray(value) ? value : [value]) : null,
            null, Gio.DBusCallFlags.NONE, 5000, this._cancellable, (bus, result) => {
                try { const response = bus.call_finish(result).deep_unpack(); if (!this._closed && done) done(response); }
                catch (error) { if (!this._closed) this.set_applet_tooltip('UsageStat: ' + error.message); }
            });
    }
    _render(state) {
        if (this._closed) return;
        this._state = state;
        this.set_applet_tooltip(state.providers.filter(p => !p.parent).map(p => p.name + ': ' + (p.error || p.text)).join('\n') || 'UsageStat — set up providers');
        try {
            const foreground = this.actor.get_theme_node().get_foreground_color();
            const vertical = [St.Side.LEFT, St.Side.RIGHT].includes(this._orientation);
            const light = foreground.red + foreground.green + foreground.blue < 382;
            const key = vertical ? light ? 'panelImageVerticalLight' : 'panelImageVertical' : light ? 'panelImageLight' : 'panelImage';
            const path = state[key + 'Png'] || state[key];
            const scale = this._theme.scale_factor;
            const thickness = vertical ? this.panel.actor.width : this.panel.actor.height;
            const size = Math.round(Math.max(16, Math.min(vertical ? 40 : 28, thickness / scale - 4)) * scale);
            const pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_scale(path, vertical ? size : -1, vertical ? -1 : size, true);
            const content = new Clutter.Image();
            content.set_data(pixbuf.get_pixels(), pixbuf.get_has_alpha() ? Cogl.PixelFormat.RGBA_8888 : Cogl.PixelFormat.RGB_888,
                pixbuf.width, pixbuf.height, pixbuf.rowstride);
            this._image.set_content(content);
            this._image.set_size(pixbuf.width, pixbuf.height);
        } catch (error) { this.set_applet_tooltip('UsageStat: ' + error.message); }
    }
    on_applet_clicked() {
        const [x, y] = this.actor.get_transformed_position();
        const [w, h] = this.actor.get_transformed_size();
        const monitor = Main.layoutManager.findMonitorForActor(this.actor);
        const workspace = (global.workspace_manager || global.screen).get_active_workspace();
        const area = workspace.get_work_area_for_monitor(monitor.index);
        const edge = { [St.Side.TOP]: 'top', [St.Side.BOTTOM]: 'bottom', [St.Side.LEFT]: 'left', [St.Side.RIGHT]: 'right' }[this._orientation];
        this._call('ToggleDetailsAt', '(ss)', ['', JSON.stringify({edge, rect: {x, y, w, h},
            work: {x: area.x, y: area.y, w: area.width, h: area.height}})]);
    }
    on_orientation_changed(orientation) {
        this._orientation = orientation;
        if (this._state) this._render(this._state);
    }
    on_panel_height_changed() {
        if (this._state) this._render(this._state);
    }
    on_applet_removed_from_panel() {
        this._closed = true;
        this._cancellable.cancel();
        this._theme.disconnect(this._scaleSignal);
        Gio.DBus.session.signal_unsubscribe(this._subscription);
        Gio.bus_unwatch_name(this._watch);
    }
}

function main(metadata, orientation, height, instanceId) {
    return new UsageStatApplet(metadata, orientation, height, instanceId);
}
